// jsmediatags/build2/MediaTagReader.js

'use strict';

const MediaFileReader = require('./MediaFileReader');

class MediaTagReader {

  constructor(mediaFileReader) {
    this._mediaFileReader = mediaFileReader;
    this._tags = null;
  }

  /**
   * Returns the byte range that needs to be loaded and fed to
   * _canReadTagFormat in order to identify if the file contains tag
   * information that can be read.
   */
  static getTagIdentifierByteRange() {
    throw new Error("Must implement");
  }

  /**
   * Given a tag identifier (read from the file byte positions speficied by
   * getTagIdentifierByteRange) this function checks if it can read the tag
   * format or not.
   */
  static canReadTagFormat(tagIdentifier) {
    throw new Error("Must implement");
  }

  setTagsToRead(tags) {
    this._tags = tags;
    return this;
  }

  read(callbacks) {
    var self = this;

    this._mediaFileReader.init({
      onSuccess: function () {
        self._loadData(self._mediaFileReader, {
          onSuccess: function () {
            var tags = self._parseData(self._mediaFileReader, self._tags);
            // TODO: destroy mediaFileReader
            callbacks.onSuccess(tags);
          },
          onError: callbacks.onError
        });
      },
      onError: callbacks.onError
    });
  }

  getShortcuts() {
    return {};
  }

  /**
   * Load the necessary bytes from the media file.
   */
  _loadData(mediaFileReader, callbacks) {
    throw new Error("Must implement _loadData function");
  }

  /**
   * Parse the loaded data to read the media tags.
   */
  _parseData(mediaFileReader, tags) {
    throw new Error("Must implement _parseData function");
  }

  _expandShortcutTags(tagsWithShortcuts) {
    if (!tagsWithShortcuts) {
      return null;
    }

    var tags = [];
    var shortcuts = this.getShortcuts();
    for (var i = 0, tagOrShortcut; tagOrShortcut = tagsWithShortcuts[i]; i++) {
      tags = tags.concat(shortcuts[tagOrShortcut] || [tagOrShortcut]);
    }

    return tags;
  }
}

module.exports = MediaTagReader;
